### ISSUE TITLE
cargo should avoid releasing jobserver tokens when it's going to need a new one

### ISSUE THREAD

**Issue by glandium** [rust-lang/cargo#7786]
When building with cargo under a GNU make jobserver, cargo races with whatever else GNU make is doing. In the case of Firefox, cargo is mostly racing with clang-c++ or g++.

Because rust build times are generally worse than C++ build times, combined with the serialized nature of some crate dependencies in Firefox, it's important for rust code to build as quickly as possible, because that's what the build ends up waiting for when there is enough parallelism. And to do that it's important for rust compilation to start as quickly as possible.

What I've observed happens in practice, is that because cargo races with the C++ compiler to get tokens, it can end up waiting for new tokens, although it was holding one just moments before for a job that just finished. And this can (and does) delay rust compilations by that time it takes to get that token.

That being said, it doesn't seem to be adding that much overhead all things considered. At least it seems to be less than the overhead of waiting for tokens at the very start of the cargo build, but there, there's really nothing that cargo can do... although I wonder if one read(50) would allow to get 50 tokens faster than 50 read(1) on the jobserver pipe... (or is it write(), I forget)

**Comment by ehuss** [comment]
cc @Mark-Simulacrum does the new `-Zjobserver-token-requests` address this request?  IIUC, tokens won't be released as aggressively as before (though they are still released if not needed, right?).

**Comment by Mark-Simulacrum** [comment]
I don't think so, no. Or at least not the core of the request. That flag (potentially named differently, though, something feels off there) will help -- currently, both Cargo and the rustc's it spawns will be fighting for jobserver tokens, which means we're more likely to release a token from a rustc thread (and then not "get" it in Cargo). With the flag, Cargo will (at least in theory) be able to directly reuse that token.

I think this is likely the best we can do unless we add a "greedy" flag or something which retains tokens for, say, 50 milliseconds, before releasing them into the general pool. That would be something you only want to do in jobserver-per-rustc mode, though.

**Comment by ehuss** [comment]
@glandium We were going through our backlog and wanted to check in if you were still interested in this? 

We had some concerns about the possibility of cargo ending up being too greedy if it held onto tokens. We were also uncertain about what the actual performance benefit would be, and were wondering if a prototype could show if this made a significant difference. In general we are leaning towards closing this issue due to the potential risks and unknowns.
